File: backend/app/routers/earnings.py
"""
Earnings Calendar API Router

Provides endpoints for:
- Earnings calendar (daily, weekly, monthly views)
- Individual stock earnings history and upcoming dates
- Earnings details with historical performance
"""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta

# Import earnings functions from fmp_fetcher
from ..fmp_fetcher import (
    fetch_earnings_calendar,
    fetch_earnings_history,
    fetch_stock_earnings_info,
    is_fmp_available,
)


router = APIRouter()


# ====== Pydantic Models ======

class EarningEvent(BaseModel):
    """Single earnings announcement event."""
    symbol: str = Field(..., description="Stock ticker symbol")
    date: str = Field(..., description="Earnings announcement date (YYYY-MM-DD)")
    time: Optional[str] = Field(None, description="Time of announcement (bmo, amc, dmh, unknown)")
    epsEstimate: Optional[float] = Field(None, description="Estimated EPS")
    epsActual: Optional[float] = Field(None, description="Actual reported EPS (if available)")
    revenueEstimate: Optional[float] = Field(None, description="Estimated revenue")
    revenueActual: Optional[float] = Field(None, description="Actual revenue (if available)")
    fiscalDateEnding: Optional[str] = Field(None, description="Fiscal period end date")


class DailyEarnings(BaseModel):
    """Earnings grouped by a single day."""
    date: str = Field(..., description="Date (YYYY-MM-DD)")
    dayOfWeek: str = Field(..., description="Day name (Monday, Tuesday, etc.)")
    earnings: List[EarningEvent] = Field(default_factory=list)
    count: int = Field(0, description="Number of earnings announcements")


class EarningsCalendarResponse(BaseModel):
    """Response for earnings calendar endpoint."""
    startDate: str = Field(..., description="Calendar start date")
    endDate: str = Field(..., description="Calendar end date")
    view: str = Field(..., description="View type (daily, weekly, monthly)")
    days: List[DailyEarnings] = Field(default_factory=list)
    totalEarnings: int = Field(0, description="Total number of earnings in period")


class HistoricalEarning(BaseModel):
    """Single historical earnings report."""
    date: str = Field(..., description="Earnings report date")
    symbol: str = Field(..., description="Stock ticker")
    epsActual: Optional[float] = Field(None, description="Actual EPS reported")
    epsEstimate: Optional[float] = Field(None, description="EPS estimate")
    epsSurprise: Optional[float] = Field(None, description="EPS surprise (actual - estimate)")
    epsSurprisePercent: Optional[float] = Field(None, description="EPS surprise percentage")
    revenue: Optional[float] = Field(None, description="Actual revenue")
    revenueEstimate: Optional[float] = Field(None, description="Estimated revenue")
    fiscalDateEnding: Optional[str] = Field(None, description="Fiscal period end")


class EarningsStats(BaseModel):
    """Statistics about historical earnings performance."""
    totalReported: int = Field(0, description="Total number of earnings reports")
    beats: int = Field(0, description="Number of times beat estimates")
    misses: int = Field(0, description="Number of times missed estimates")
    meets: int = Field(0, description="Number of times met estimates")
    beatRate: Optional[float] = Field(None, description="Percentage of beats")


class StockEarningsResponse(BaseModel):
    """Complete earnings info for a single stock."""
    symbol: str = Field(..., description="Stock ticker")
    nextEarningsDate: Optional[str] = Field(None, description="Next earnings date")
    nextEarningsTime: Optional[str] = Field(None, description="Time of next earnings (bmo/amc)")
    epsEstimate: Optional[float] = Field(None, description="Next EPS estimate")
    revenueEstimate: Optional[float] = Field(None, description="Next revenue estimate")
    history: List[HistoricalEarning] = Field(default_factory=list)
    stats: EarningsStats = Field(default_factory=EarningsStats)


class BulkEarningsRequest(BaseModel):
    """Request for earnings dates of multiple stocks."""
    tickers: List[str] = Field(..., description="List of stock tickers")


class BulkEarningsItem(BaseModel):
    """Earnings date info for a single stock."""
    symbol: str
    nextEarningsDate: Optional[str] = None
    nextEarningsTime: Optional[str] = None
    daysUntilEarnings: Optional[int] = None


class BulkEarningsResponse(BaseModel):
    """Response for bulk earnings dates."""
    earnings: List[BulkEarningsItem] = Field(default_factory=list)


# ====== Helper Functions ======

def get_day_of_week(date_str: str) -> str:
    """Get day of week name from date string."""
    try:
        date = datetime.strptime(date_str, "%Y-%m-%d")
        return date.strftime("%A")
    except:
        return "Unknown"


def organize_earnings_by_day(earnings: List[Dict], start_date: str, end_date: str) -> List[DailyEarnings]:
    """Organize earnings events by day."""
    # Create a dict to group earnings by date
    earnings_by_date: Dict[str, List[EarningEvent]] = {}
    
    for earning in earnings:
        date = earning.get("date")
        if date:
            if date not in earnings_by_date:
                earnings_by_date[date] = []
            earnings_by_date[date].append(EarningEvent(**earning))
    
    # Generate all dates in range and create DailyEarnings objects
    days = []
    current = datetime.strptime(start_date, "%Y-%m-%d")
    end = datetime.strptime(end_date, "%Y-%m-%d")
    
    while current <= end:
        date_str = current.strftime("%Y-%m-%d")
        day_earnings = earnings_by_date.get(date_str, [])
        
        days.append(DailyEarnings(
            date=date_str,
            dayOfWeek=current.strftime("%A"),
            earnings=day_earnings,
            count=len(day_earnings)
        ))
        current += timedelta(days=1)
    
    return days


# ====== API Endpoints ======

@router.get("/earnings/calendar", response_model=EarningsCalendarResponse)
async def get_earnings_calendar(
    view: str = Query("weekly", description="View type: daily, weekly, monthly"),
    from_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    to_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
):
    """
    Get earnings calendar for a specified period.
    
    Views:
    - daily: Returns next 7 days
    - weekly: Returns current week and next week (14 days)
    - monthly: Returns next 30 days
    """
    if not is_fmp_available():
        raise HTTPException(
            status_code=503,
            detail="Earnings calendar not available. Please configure FMP_API_KEY."
        )
    
    today = datetime.now()
    
    # Determine date range based on view
    if from_date and to_date:
        start = from_date
        end = to_date
    elif view == "daily":
        start = today.strftime("%Y-%m-%d")
        end = (today + timedelta(days=7)).strftime("%Y-%m-%d")
    elif view == "monthly":
        start = today.strftime("%Y-%m-%d")
        end = (today + timedelta(days=30)).strftime("%Y-%m-%d")
    else:  # weekly (default)
        # Start from beginning of current week (Monday)
        days_since_monday = today.weekday()
        week_start = today - timedelta(days=days_since_monday)
        start = week_start.strftime("%Y-%m-%d")
        end = (week_start + timedelta(days=13)).strftime("%Y-%m-%d")  # 2 weeks
    
    try:
        # Fetch earnings from FMP
        earnings = fetch_earnings_calendar(start, end)
        
        # Organize by day
        days = organize_earnings_by_day(earnings, start, end)
        
        return EarningsCalendarResponse(
            startDate=start,
            endDate=end,
            view=view,
            days=days,
            totalEarnings=len(earnings)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching earnings calendar: {str(e)}")


@router.get("/earnings/{ticker}", response_model=StockEarningsResponse)
async def get_stock_earnings(ticker: str):
    """
    Get complete earnings information for a specific stock.
    
    Includes:
    - Next earnings date and time
    - EPS and revenue estimates
    - Historical earnings with beat/miss analysis
    - Overall earnings performance statistics
    """
    if not is_fmp_available():
        raise HTTPException(
            status_code=503,
            detail="Earnings data not available. Please configure FMP_API_KEY."
        )
    
    try:
        earnings_info = fetch_stock_earnings_info(ticker)
        
        if not earnings_info:
            raise HTTPException(status_code=404, detail=f"No earnings data found for {ticker}")
        
        # Convert history to HistoricalEarning models
        history = [HistoricalEarning(**h) for h in earnings_info.get("history", [])]
        
        # Create stats model
        stats_data = earnings_info.get("stats", {})
        stats = EarningsStats(
            totalReported=stats_data.get("totalReported", 0),
            beats=stats_data.get("beats", 0),
            misses=stats_data.get("misses", 0),
            meets=stats_data.get("meets", 0),
            beatRate=stats_data.get("beatRate"),
        )
        
        return StockEarningsResponse(
            symbol=earnings_info.get("symbol", ticker.upper()),
            nextEarningsDate=earnings_info.get("nextEarningsDate"),
            nextEarningsTime=earnings_info.get("nextEarningsTime"),
            epsEstimate=earnings_info.get("epsEstimate"),
            revenueEstimate=earnings_info.get("revenueEstimate"),
            history=history,
            stats=stats,
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching earnings for {ticker}: {str(e)}")


@router.get("/earnings/{ticker}/history")
async def get_earnings_history(
    ticker: str,
    limit: int = Query(20, ge=1, le=50, description="Number of historical records")
):
    """
    Get historical earnings for a specific stock.
    """
    if not is_fmp_available():
        raise HTTPException(
            status_code=503,
            detail="Earnings data not available. Please configure FMP_API_KEY."
        )
    
    try:
        history = fetch_earnings_history(ticker, limit)
        return {"ticker": ticker.upper(), "history": history, "count": len(history)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching earnings history: {str(e)}")


@router.post("/earnings/bulk", response_model=BulkEarningsResponse)
async def get_bulk_earnings_dates(request: BulkEarningsRequest):
    """
    Get next earnings dates for multiple stocks at once.
    Useful for portfolio/watchlist views.
    """
    if not is_fmp_available():
        raise HTTPException(
            status_code=503,
            detail="Earnings data not available. Please configure FMP_API_KEY."
        )
    
    if len(request.tickers) > 50:
        raise HTTPException(
            status_code=400,
            detail="Maximum 50 tickers allowed per request"
        )
    
    today = datetime.now()
    today_str = today.strftime("%Y-%m-%d")
    future_str = (today + timedelta(days=120)).strftime("%Y-%m-%d")
    
    try:
        # Fetch calendar for the range
        calendar = fetch_earnings_calendar(today_str, future_str)
        
        # Create lookup by symbol
        earnings_by_symbol: Dict[str, Dict] = {}
        for earning in calendar:
            symbol = earning.get("symbol", "").upper()
            if symbol and symbol not in earnings_by_symbol:
                earnings_by_symbol[symbol] = earning
        
        # Build response
        results = []
        for ticker in request.tickers:
            ticker_upper = ticker.upper()
            earning_data = earnings_by_symbol.get(ticker_upper)
            
            days_until = None
            if earning_data and earning_data.get("date"):
                try:
                    earnings_date = datetime.strptime(earning_data["date"], "%Y-%m-%d")
                    days_until = (earnings_date - today).days
                except:
                    pass
            
            results.append(BulkEarningsItem(
                symbol=ticker_upper,
                nextEarningsDate=earning_data.get("date") if earning_data else None,
                nextEarningsTime=earning_data.get("time") if earning_data else None,
                daysUntilEarnings=days_until,
            ))
        
        return BulkEarningsResponse(earnings=results)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching bulk earnings: {str(e)}")

File: frontend/lib/api/earnings.ts
/**
 * Earnings Calendar API services.
 */

import { api } from './client';

// ====== Types ======

export interface EarningEvent {
    symbol: string;
    date: string;
    time?: string | null;  // "bmo" (before market open), "amc" (after market close), "dmh" (during market hours)
    epsEstimate?: number | null;
    epsActual?: number | null;
    revenueEstimate?: number | null;
    revenueActual?: number | null;
    fiscalDateEnding?: string | null;
}

export interface DailyEarnings {
    date: string;
    dayOfWeek: string;
    earnings: EarningEvent[];
    count: number;
}

export interface EarningsCalendarResponse {
    startDate: string;
    endDate: string;
    view: string;
    days: DailyEarnings[];
    totalEarnings: number;
}

export interface HistoricalEarning {
    date: string;
    symbol: string;
    epsActual?: number | null;
    epsEstimate?: number | null;
    epsSurprise?: number | null;
    epsSurprisePercent?: number | null;
    revenue?: number | null;
    revenueEstimate?: number | null;
    fiscalDateEnding?: string | null;
}

export interface EarningsStats {
    totalReported: number;
    beats: number;
    misses: number;
    meets: number;
    beatRate?: number | null;
}

export interface StockEarningsResponse {
    symbol: string;
    nextEarningsDate?: string | null;
    nextEarningsTime?: string | null;
    epsEstimate?: number | null;
    revenueEstimate?: number | null;
    history: HistoricalEarning[];
    stats: EarningsStats;
}

export interface BulkEarningsItem {
    symbol: string;
    nextEarningsDate?: string | null;
    nextEarningsTime?: string | null;
    daysUntilEarnings?: number | null;
}

export interface BulkEarningsResponse {
    earnings: BulkEarningsItem[];
}

// ====== API Functions ======

export type CalendarView = 'daily' | 'weekly' | 'monthly';

/**
 * Get earnings calendar for a specified period.
 */
export async function getEarningsCalendar(
    view: CalendarView = 'weekly',
    fromDate?: string,
    toDate?: string
): Promise<EarningsCalendarResponse> {
    const params = new URLSearchParams({ view });
    if (fromDate) params.append('from_date', fromDate);
    if (toDate) params.append('to_date', toDate);

    const response = await fetch(`http://localhost:8100/api/earnings/calendar?${params}`);
    if (!response.ok) {
        throw new Error('Failed to fetch earnings calendar');
    }
    return response.json();
}

/**
 * Get complete earnings info for a specific stock.
 */
export async function getStockEarnings(ticker: string): Promise<StockEarningsResponse> {
    const response = await fetch(`http://localhost:8100/api/earnings/${ticker}`);
    if (!response.ok) {
        throw new Error(`Failed to fetch earnings for ${ticker}`);
    }
    return response.json();
}

/**
 * Get historical earnings for a specific stock.
 */
export async function getEarningsHistory(
    ticker: string,
    limit: number = 20
): Promise<{ ticker: string; history: HistoricalEarning[]; count: number }> {
    const response = await fetch(`http://localhost:8100/api/earnings/${ticker}/history?limit=${limit}`);
    if (!response.ok) {
        throw new Error(`Failed to fetch earnings history for ${ticker}`);
    }
    return response.json();
}

/**
 * Get next earnings dates for multiple stocks at once.
 * Useful for portfolio/watchlist views.
 */
export async function getBulkEarningsDates(tickers: string[]): Promise<BulkEarningsResponse> {
    const response = await fetch('http://localhost:8100/api/earnings/bulk', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ tickers }),
    });
    if (!response.ok) {
        throw new Error('Failed to fetch bulk earnings dates');
    }
    return response.json();
}

// ====== Helper Functions ======

/**
 * Format earnings time code to human-readable string.
 */
export function formatEarningsTime(time?: string | null): string {
    if (!time) return '';

    switch (time.toLowerCase()) {
        case 'bmo':
            return 'Before Market Open';
        case 'amc':
            return 'After Market Close';
        case 'dmh':
            return 'During Market Hours';
        default:
            return time;
    }
}

/**
 * Format earnings time to short code.
 */
export function formatEarningsTimeShort(time?: string | null): string {
    if (!time) return '';

    switch (time.toLowerCase()) {
        case 'bmo':
            return 'Pre-Market';
        case 'amc':
            return 'After Hours';
        case 'dmh':
            return 'Market Hours';
        default:
            return time.toUpperCase();
    }
}

/**
 * Calculate days until earnings.
 */
export function getDaysUntilEarnings(dateStr?: string | null): number | null {
    if (!dateStr) return null;

    try {
        const earningsDate = new Date(dateStr);
        const today = new Date();
        today.setHours(0, 0, 0, 0);
        earningsDate.setHours(0, 0, 0, 0);

        const diffTime = earningsDate.getTime() - today.getTime();
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

        return diffDays;
    } catch {
        return null;
    }
}

/**
 * Format days until earnings to human-readable string.
 */
export function formatDaysUntilEarnings(days: number | null): string {
    if (days === null) return 'Unknown';
    if (days === 0) return 'Today';
    if (days === 1) return 'Tomorrow';
    if (days < 0) return `${Math.abs(days)} days ago`;
    if (days <= 7) return `In ${days} days`;
    if (days <= 14) return 'Next week';
    if (days <= 30) return `In ${Math.ceil(days / 7)} weeks`;
    return `In ${Math.ceil(days / 30)} months`;
}

/**
 * Get urgency color class based on days until earnings.
 */
export function getEarningsUrgencyClass(days: number | null): string {
    if (days === null) return 'text-gray-400';
    if (days <= 0) return 'text-red-600 font-bold';
    if (days <= 3) return 'text-orange-600 font-semibold';
    if (days <= 7) return 'text-yellow-600';
    return 'text-gray-600';
}

/**
 * Format surprise percentage with color indication.
 */
export function getSurpriseInfo(percent?: number | null): {
    text: string;
    colorClass: string;
    isBeat: boolean | null;
} {
    if (percent === null || percent === undefined) {
        return { text: '-', colorClass: 'text-gray-400', isBeat: null };
    }

    const absPercent = Math.abs(percent);
    const formattedPercent = absPercent.toFixed(2);

    if (percent > 0.5) {
        return {
            text: `Beat +${formattedPercent}%`,
            colorClass: 'text-green-600',
            isBeat: true,
        };
    } else if (percent < -0.5) {
        return {
            text: `Miss -${formattedPercent}%`,
            colorClass: 'text-red-600',
            isBeat: false,
        };
    } else {
        return {
            text: 'Met',
            colorClass: 'text-gray-600',
            isBeat: null,
        };
    }
}
